src/fhirtypes/common.ts

```diff
@@ -32,15 +32,23 @@ export function setPropertyOnInstance(
     // instance property here
     let current: any = instance;
     for (const [i, pathPart] of pathParts.entries()) {
-      const key = pathPart.base;
+      // When a primitive has child elements, a _ is appended to the name of the primitive
+      // According to https://www.hl7.org/fhir/json.html#primitive
+      const key =
+        pathPart.primitive && i < pathParts.length - 1 ? `_${pathPart.base}` : pathPart.base;
       // If this part of the path indexes into an array, the index will be the last bracket
       const index = getArrayIndex(pathPart);
       if (index != null) {
         // If the array doesn't exist, create it
         if (current[key] == null) current[key] = [];
         // If the index doesn't exist in the array, add it and lesser indices
-        if (index >= current[key].length) {
-          current[key][index] = {};
+        // Empty elements should be null, not undefined, according to https://www.hl7.org/fhir/json.html#primitive
+        let j = current[key].length;
+        for (; j < index; j++) {
+          current[key].push(null);
+        }
+        if (j === index) {
+          current[key].push({});
```

Comment:
Note that in a previous PR, this loop was changed to
```
if (index >= current[key].length) {
          current[key][index] = {};
}
```
to be more compact. However, I had to revert back to using this loop, because the more compact syntax fills the array with `undefined`, while according to this documentation, https://www.hl7.org/fhir/json.html#primitive, the array needs to be filled with `null`.

Reply:
Alas, my clever trick! But any solution that involves appending a pre-filled `null` array of the appropriate size will probably be harder to understand, so this is completely fine with me.